src/cmap/auth/mongodb_oidc/automated_callback_workflow.ts

```diff
@@ -0,0 +1,61 @@
+import { type Document } from 'bson';
+
+import { MongoMissingCredentialsError } from '../../../error';
+import { type Connection } from '../../connection';
+import { type AuthMechanismProperties, type MongoCredentials } from '../mongo_credentials';
+import { type OIDCCallbackFunction } from '../mongodb_oidc';
+import { CallbackWorkflow } from './callback_workflow';
+import { type TokenCache, type TokenEntry } from './token_cache';
+
+const NO_CALLBACK = 'No OIDC_CALLBACK provided for callback workflow.';
+
+/**
+ * Class implementing behaviour for the non human callback workflow.
+ * @internal
+ */
+export class AutomatedCallbackWorkflow extends CallbackWorkflow {
+  /**
+   * Execute the OIDC callback workflow.
+   */
+  async execute(
+    connection: Connection,
+    credentials: MongoCredentials,
+    cache?: TokenCache
+  ): Promise<Document> {
+    const callback = getCallback(credentials.mechanismProperties);
+    // If there is a cached access token, try to authenticate with it. If
+    // authentication fails with an Authentication error (18),
+    // invalidate the access token, fetch a new access token, and try
+    // to authenticate again.
+    // If the server fails for any other reason, do not clear the cache.
+    let tokenEntry: TokenEntry;
+    if (cache?.hasToken()) {
```

Comment:
I don't feel strongly, but this is a direct port of the algorithm defined in the spec: https://github.com/mongodb/specifications/blob/master/source/auth/auth.md#two-step

Cross-referencing an implementation with the spec is easier if they're line-for-line the same.

Reply:
Sure but that's a pedantic degree of code formatting that may not last. I agree with you on the point of cross-referencing and don't mind leaving it this way. However, if it has no functional purpose and someone is driving by here, their instinct may be to collapse this to a `const` because TS and eslint encourage it and so engineers gravitate to it, and would they be wrong to? If you, I, or Durran are not there to review it then there's nothing stopping it from changing. 

So, perhaps, if we wanna keep it looking like the pseudo-code, let us mention it in a comment or "normalize" it to ward off those kinds of known drive-bys